계산기 [Step1] 비모

bubblecocoa opened the pull request:
@JoSH0318 
안녕하세요 밈!
계산기 프로젝트 Step1 PR 보냅니다!
2주간 잘 부탁드립니다 🙇

---

### 고민했던 점
#### 1️⃣ 시간복잡도에 따른 Queue 구현
> 배열의 중간에 값 추가﹒제거하는 기능은 배제한 `Queue` 형태만 고려했습니다. 
1. 일반 Array 이용
    `enqueue`시 사용하는 메서드 [append(_:)](https://developer.apple.com/documentation/swift/array/append(_:)-1ytnt)는 맨 뒤에 추가하는 작업으로 시간 복잡도가 `O(1)`이지만,  `dequeue`시 사용하는 메서드 [removefirst()](https://developer.apple.com/documentation/swift/array/removefirst())는 맨 앞 요소 제거 완료 후 나머지 요소를 하나씩 앞으로 당겨오는 작업으로 인해 시간 복잡도가 `O(n)`입니다.
2. Array에 포인터 생성
    1번의 `dequeue`시 발생하는 [오버헤드](https://ko.wikipedia.org/wiki/%EC%98%A4%EB%B2%84%ED%97%A4%EB%93%9C)를 해결하기 위해 포인터 역할을 할 수 있는 변수를 선언(name e.g. `head`, `point`, `index` 등)합니다. `dequeue`시 현재 포인터에 해당하는 값을 반환해주고, 포인터를 뒤로 하나 이동합니다. 실제 값은 제거하지 않고 포인터만 이동하는 것으로 시간 복잡도는 `O(1)`입니다. 하지만 실제 배열에서 제거된 것은 아니기 때문에 어떤 조건을 통해 [removefirst(_:)](https://developer.apple.com/documentation/swift/array/removefirst(_:))를 사용해야 하는데, 이것의 시간 복잡도는 `O(n)`입니다.
3. 단방향 LinkedList
    `head`값을 두어 첫번째 노드가 무엇인지 알고있어야 합니다. `next`값을 두어 다음 노드를 알 수 있습니다. `next`가 `nil`일 경우 마지막 노드입니다. 이 설정들을 통해 `enqueue`시 `head`부터 시작해 `node?.next`가 `nil`인 경우까지 순회를 통해 찾은 후 `node?.next`에 요소를 추가합니다. 이 작업은 단일 순회를 통해 이루어져 시간 복잡도가 `O(n)`입니다. `dequeue`시 현재 `head` 노드를 반환 후 제거, 다음 노드를 `head`로 지정합니다. 이 작업의 시간 복잡도는 `O(1)`입니다.
4. Double Stack 이용
    `enqueue`용 `stack`, `dequeue`용 `stack` 두가지를 이용해 구현하는 방식입니다. `enqueue`시 `enqueueStack`에 [append(_:)](https://developer.apple.com/documentation/swift/array/append(_:)-1ytnt)를 이용해 값을 추가하고, `dequeue`시 [removelast()](https://developer.apple.com/documentation/swift/array/removelast())를 이용해 값을 제거합니다. `dequeue`시 `dequeueStack`가 비어있는 상태라면 `enqueueStack`을 [reversed()](https://developer.apple.com/documentation/swift/array/reversed())한 값을 전달받고 `enqueueStack`의 값을 [removeall()](https://developer.apple.com/documentation/swift/array/removeall(keepingcapacity:)-1er5) 합니다. 이 과정에서 `reversed()`와 `removeAll()`는 모두 `O(n)`이기 때문에 `O(2n)`으로 볼 수 있지만 수치상 큰 의미가 없어 `O(n)`으로 볼 수 있습니다.

- 요약
    |시간 복잡도|enqeue|dequeue|
    |:-:|:-:|:-:|
    |**Array**|O(1)|O(n)|
    |**Pointer Array**|O(1)|O(1)<br>실제 값 제거 시 O(n)|
    |**단방향 LinkedList**|O(n)|O(1)|
    |**Double Stack**|O(1)|O(1)<br>dequeueStack 비어있으면 O(n)|

- 결론
    시간 복잡도가 낮은 `Pointer Array`와 `Double Stack` 둘 중에 고민했습니다. `Pointer Array`에서 `dequeue`시 포인터에 해당하는 값을 `nil`로 변경하고 포인터를 이동하는 방식도 매력적이었으나, 계산기 특성상 `queue`에 쌓일 값이 방대하지 않을 것 같아 배열이 `nil`값을 가지고 있지 않을 `Double Stack`을 선택했습니다.

---

### 조언을 구하는 점
#### 1️⃣ TDD: 테스트 코드가 에러인 상태에서 커밋?
TDD는 Red: 테스트가 실패하는 코드 작성 -> Green: 테스트가 성공하는 코드 작성 -> Refactor: 코드 리팩토링 순서대로 개발하는 것으로 알고 있습니다. 그런데 커밋 단위를 고민하다보니 테스트코드와, 실제 동작 코드를 함께 커밋 하는 것이 이상한듯 해 분리하여 커밋 했습니다. 이런 방식의 커밋은 처음 해보는데 어떤 것이 권장 사항인지 궁금합니다.<br> 추가로 유닛테스트를 제대로 작성한것이 맞는지 궁금합니다.

#### 2️⃣ 컨벤션: 연산프로퍼티의 개행
```swift
struct CalculatorItemQueue<Element: CalculateItem> {
    ...
    var isEmpty: Bool {
        return enqueueStack.isEmpty && dequeueStack.isEmpty
    }
    var count: Int {
        return enqueueStack.count + dequeueStack.count
    }
    ...
}
```
만약 연산 프로퍼티가 아니었다면 개행없이 쭉 썼을테지만 클로저가 닫힌 후 개행이 없는 부분이 어색해보였습니다. 어떻게 개행하는게 보편적인 방법인지 궁금합니다.

bubblecocoa committed:
rename: MVC 구조로 변경

bubblecocoa committed:
feat: CalculatorItemQueue, CalculateItem 생성

- 프로토콜 CalculateItem 생성
- 클래스 CalculatorItemQueue 생성, 프로토콜 CalculateItem를 준수하는 저장 프로퍼티 calculateItems 선언

bubblecocoa committed:
refactor: CalculatorItemQueue가 CalculateItem을 준수하는 Element를 받도록 변경

bubblecocoa committed:
test: CalculatorItemQueueTests 생성

- CalculatorItemQueue가 Double을 받도록 지정

bubblecocoa committed:
feat: Double이 CalculateItem프로토콜을 채택하는 extension 추가

bubblecocoa committed:
test: CalculatorItemQueue의 isEmpty 테스트 케이스 추가

bubblecocoa committed:
refactor: CalculatorItemQueue를 더블스택으로 구현하기 위해 변경

bubblecocoa committed:
feat: CalculatorItemQueue에 읽기 전용 연산 프로퍼티 isEmpty 추가

bubblecocoa committed:
test: CalculatorItemQueue의 isEmpty 테스트 케이스 추가

- 큐에 값이 있을때 isEmpty를 조회하면 false인지 확인

bubblecocoa committed:
feat: CalculatorItemQueue에 enqueue 메서드 추가

bubblecocoa committed:
test: CalculatorItemQueue의 dequeue 테스트 케이스 추가

- 큐의 선입선출이 정상 동작하는지 확인하는 테스트

bubblecocoa committed:
feat: CalculatorItemQueue에 dequeue 메서드 추가

- 더블스택을 이용해 dequeue

bubblecocoa committed:
test: CalculatorItemQueue의 count 테스트 케이스 추가

bubblecocoa committed:
feat: CalculatorItemQueue에 읽기 전용 연산 프로퍼티 count 추가

- count는 enqueue와 dequeue의 count 합을 리턴한다

bubblecocoa committed:
test: CalculatorItemQueue의 peek 테스트 케이스 추가

bubblecocoa committed:
feat: CalculatorItemQueue에 peek 메서드 추가

- dequeue하지 않고 다음 out때 무엇이 나올지 확인하는 메서드

bubblecocoa committed:
test: CalculatorItemQueue의 dequeue 테스트 케이스 추가

bubblecocoa committed:
refactor: CalculatorItemQueue의 dequeue메서드가 nil을 반환할 수 있도록 변경

- removeLast는 Self.Element / popLast는 Self.Element?

bubblecocoa committed:
test: CalculatorItemQueue의 removeAll 테스트 케이스 추가

bubblecocoa committed:
feat: CalculatorItemQueue에 removeAll 메서드 추가

bubblecocoa committed:
style: 컨벤션에 맞게 코드 위치 변경

bubblecocoa committed:
docs: step1 클래스 다이어그램 추가

bubblecocoa committed:
docs: step 1 클래스 다이어그램 변경

JoSH0318 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,39 @@
+        enqueueStack.append(element)
+    }
+    
+    mutating func dequeue() -> Element? {
+        if dequeueStack.isEmpty {
+            dequeueStack = enqueueStack.reversed()
+            enqueueStack.removeAll()
+        }
+        
+        return dequeueStack.popLast()
+    }
```
만약 `enqueueStack` `dequeueStack` 둘다 비어있는 상황에서 `dequeue`를 호출했다 가정하면
> `dequeueStack.isEmpty`가 ture
-> 비어있는 `enqueueStack`을 `reversed()` 
-> `dequeueStack`에 할당 
-> 비어있는 `enqueueStack`을 `removeAll()` 
-> 반환
순으로 흘러가겠네요!

조금 비효율적인 동작을 하는 것 같습니다! 만들어주신 isEmpty를 사용하면 더 좋은 결과가 나올 것 같은데 
어떻게 생각하시나요?

bubblecocoa replied:
`dequeue`가 비어있는 상황에만 집중해서 비효율적인 동작을 놓쳤네요!
기존 읽기 전용 연산 프로퍼티 `isEmpty`를 활용하는 코드로 변경했습니다!

JoSH0318 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,39 @@
+        return dequeueStack.popLast()
+    }
+    
+    func peek() -> Element? {
+        !enqueueStack.isEmpty ? enqueueStack.first : dequeueStack.last
+    }
```
dequeue가 아닌 첫번째 값을 조회하는 기능!!! 너무 좋은 것 같습니다. 👍
다만 개인적으로 `!enqueueStack.isEmpty`은 부정의 부정이라 가독성이 조금 떨어지는 경향이 있는 것 같아요.
때문에 `!`는 조건문에서 사용하지 않게 되는 것 같습니다. 
물론 별다른 방법이 없다면 사용하는게 맞겠지만 현재는 방법이 있을 것 같습니다!

bubblecocoa replied:
'리턴된 `Bool`의 반대'보다 '리턴된 `Bool`'이라고 생각하는 것이 가독성이 좋아지겠네요!
`enqueueStack.isEmpty ? dequeueStack.last : enqueueStack.first`로 변경했습니다!

JoSH0318 commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,115 @@
+final class CalculatorItemQueueTests: XCTestCase {
+    var sut: CalculatorItemQueue<Double>!
+    
+    override func setUpWithError() throws {
+        sut = CalculatorItemQueue()
+    }
+
+    override func tearDownWithError() throws {
+        sut = nil
+    }
```
super를 호출하지 않고 있는데 이유가 있을까요? 🤔

bubblecocoa replied:
별도 의도가 있던것은 아니고 누락되었습니다!
`override`로 내용을 덮어써도 문제가 없을 수 있지만 `super.setUpWithError()`는 초기 테스트 초기 상태 설정, 리소스 할당의 역할을 하는 것으로 알고 있습니다. 내부 구현 코드를 알 수 없기 때문에 해당 코드는 누락하지 않고 호출해 주는 것이 좋을 것으로 판단됩니다!

JoSH0318 commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,115 @@
+    
+    func test_count호출시_현재_큐에_쌓인_수를_리턴한다() {
+        // given
+        sut.enqueue(Double.random(in: -10...10))
+        sut.enqueue(Double.random(in: -10...10))
+        sut.enqueue(Double.random(in: -10...10))
+        sut.enqueue(Double.random(in: -10...10))
```
이 부분이 조금 고민입니다. 😀
테스트는 예상할 수 있는 값으로 진행해야하지만 현재는 예상하지 않아도 되는 상황이네요.🥹
하지만 이 테스트에서 random한 Double이 주는 의미를 모르겠어요.😂
이 부분에 대해서 의견을 주시면 감사하겠습니다!

bubblecocoa replied:
`Double.random`값을 사용한 테스트의 경우 '어떤 `Double` 값이 와도 상관없다.' 라는 의미로 `given`을 주었습니다!
테스트가 좀 더 통제된 상황에서 진행되는게 좋을까요?

JoSH0318 replied:
될 수 있다면 모든 조건이 통제된 상황이 좋다 라는 생각입니다. 🥹
하지만 비모가 말씀하신대로 '어떤 Double 값이 와도 상관없다.'라는 의미가 담겼다면 이 또한 정답이겠네요!
좋습니다 👍

JoSH0318 commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,115 @@
+        XCTAssertEqual(result, expectedResult)
+    }
+    
+    func test_dequeue호출시_첫_enqueue_값을_리턴한다() {
```
개인적으로 네이밍이 조금더 명확했으면 좋겠습니다.
`어떠한 조건에서` `어떠한 상황에 의해` `어떠한 값을 예상하는지`를 명시해야 좋다고 생각합니다.
예를 들어 `test_dequeue_1_2_3_enqueue시_dequeue하면_1을_반환한다` 처럼 말이죠!
dequeue에 대한 test임을 밝히고, given/when/then에 맞춰 네이밍을 하면 좋을 것 같아요!
한가지의 예시일 뿐이니, 비모의 생각에 맞춰 수정해보시면 좋을 것 같아요!

bubblecocoa replied:
좀 더 구체적인 값인 `1.0`, `2.0`, `3.0`을  테스트 메서드명으로 주고 싶었으나 `.`을 메서드명에 포함시킬 수 없어 밈이 주신 예와 다르지만 `given/when/then`에 맞춰 네이밍을 변경 해보았습니다!

JoSH0318 reviewed (commented):
안녕하세요 🫶 비모 🔋

이유있는 코드를 작성해주신 점👍 너무너무 좋습니다.
천천히 pr을 하신만큼 코드도 깔끔하다는 생각이 들어요!
역시 비모 👍

몇가지 코멘트 남겼습니다!

---

### 답변
1️⃣ TDD: 테스트 코드가 에러인 상태에서 커밋?
Q: TDD는 Red: 테스트가 실패하는 코드 작성 -> Green: 테스트가 성공하는 코드 작성 -> Refactor: 코드 리팩토링 순서대로 개발하는 것으로 알고 있습니다. 그런데 커밋 단위를 고민하다보니 테스트코드와, 실제 동작 코드를 함께 커밋 하는 것이 이상한듯 해 분리하여 커밋 했습니다. 이런 방식의 커밋은 처음 해보는데 어떤 것이 권장 사항인지 궁금합니다.
추가로 유닛테스트를 제대로 작성한것이 맞는지 궁금합니다.

> A: 저는 지금과 같은 commit 형태도 좋다고 생각합니다. test 코드의 에러는 app에 영향을 주지 않으니까요! 또한, 예상한 / 의도된 실패하는 코드이기 때문에 의미 있는 코드이고 commit이라 생각합니다. 어쩌면 기능 구현의 첫 시작을 기록하는 의미도 있겠네요!

2️⃣ 컨벤션: 연산프로퍼티의 개행
struct CalculatorItemQueue<Element: CalculateItem> {
    ...
    var isEmpty: Bool {
        return enqueueStack.isEmpty && dequeueStack.isEmpty
    }
    var count: Int {
        return enqueueStack.count + dequeueStack.count
    }
    ...
}

Q: 만약 연산 프로퍼티가 아니었다면 개행없이 쭉 썼을테지만 클로저가 닫힌 후 개행이 없는 부분이 어색해보였습니다. 어떻게 개행하는게 보편적인 방법인지 궁금합니다.

> A: 보편적인 방법은 없는거 아시죠!?☺️ 예상하셨겠지만, 개인/팀의 컨벤션에 맞춰 일관성 있게 작성되면 문제 없다고 생각해요. 다만 개인적인 견해로는 연산 프로퍼티는 프로퍼티이지만 func의 성격을 띄는 연산 과정이 있다고 봐요. 따라서 개행을 통해 분리하는 게 좋다고 생각합니다. 아주 주관적인 의견입니다.☺️
그렇다면!!! 연산 프로퍼티와 함수의 차이를 찾아보는 시간을 가져도 좋겠네요! 어떠한 차이점이 있을까요?

bubblecocoa committed:
refactor: CalculatorItemQueue의 dequeue 메서드 if 조건 변경

- 기존: dequeueStack이 비었는지 확인
- 변경: isEmpty 연산 프로퍼티 이용해 enqueueStack, dequeueStack이 모두 비었는지 확인

bubblecocoa committed:
refactor: CalculatorItemQueue의 3항연산자 조건 변경

- enqueueStack.isEmpty -> !enqueueStack.isEmpty

bubblecocoa committed:
test: setUpWithError()시 super.setUpWithError() 호출 누락 수정

bubblecocoa committed:
test: 테스트 메서드명 변경

bubblecocoa committed:
test: 테스트 메서드명 변경

bubblecocoa committed:
style: 읽기 전용 연산 프로퍼티에 개행 추가

bubblecocoa commented:
> A: 보편적인 방법은 없는거 아시죠!?☺️ 예상하셨겠지만, 개인/팀의 컨벤션에 맞춰 일관성 있게 작성되면 문제 없다고 생각해요. 다만 개인적인 견해로는 연산 프로퍼티는 프로퍼티이지만 func의 성격을 띄는 연산 과정이 있다고 봐요. 따라서 개행을 통해 분리하는 게 좋다고 생각합니다. 아주 주관적인 의견입니다.☺️
그렇다면!!! 연산 프로퍼티와 함수의 차이를 찾아보는 시간을 가져도 좋겠네요! 어떠한 차이점이 있을까요?

연산프로퍼티의 개행을 추가했습니다! 뭔가 마음이 편하네요. ☺️

연산 프로퍼티와 함수의 차이입니다!
- 연산프로퍼티는 파라미터를 받을 수 없습니다. 반면 함수는 여러 파라미터를 받아 값을 처리할 수 있습니다.
- 연산프로퍼티는 읽기 전용이거나 읽고 쓸 수 있어야 합니다. 이것은 연산프로퍼티 조회시 반드시 호출시 있다는 것을 의미합니다. 반면 함수는 특정 행위를 수행만 하고 리턴이 없을 수 있습니다.

JoSH0318 commented:
요구사항 만족하셨습니다 ! 특별히 언급할 부분이 없어 merge 하도록 하겠습니다.
